node/cmd/ccq/utils.go

```diff
@@ -260,19 +272,24 @@ func validateRequest(logger *zap.Logger, env common.Environment, perms Permissio
 			}
 		case *query.EthCallByTimestampQueryRequest:
 			for _, callData := range q.CallData {
-				if q.TargetTimestamp == 0 {
-					logger.Debug("eth call by timestamp must have a non-zero timestamp", zap.String("userName", permsForUser.userName))
-					return http.StatusBadRequest, fmt.Errorf("eth call by timestamp must have a non-zero timestamp")
+				contractAddress, err := vaa.BytesToAddress(callData.To)
+				if err != nil {
+					logger.Debug("failed to parse contract address", zap.String("userName", permsForUser.userName), zap.String("contract", hex.EncodeToString(callData.To)), zap.Error(err))
+					return http.StatusBadRequest, fmt.Errorf("failed to parse contract address: %w", err)
 				}
-				// TODO: For now, the block hints are required!
-				if q.TargetBlockIdHint == "" {
-					logger.Debug("eth call by timestamp must have the target block hint", zap.String("userName", permsForUser.userName))
-					return http.StatusBadRequest, fmt.Errorf("eth call by timestamp must have the target block hint")
+				if len(callData.Data) < ETH_CALL_SIG_LENGTH {
+					logger.Debug("eth call data must be at least four bytes", zap.String("userName", permsForUser.userName), zap.String("data", hex.EncodeToString(callData.Data)))
+					return http.StatusBadRequest, fmt.Errorf("eth call data must be at least four bytes")
 				}
-				if q.FollowingBlockIdHint == "" {
-					logger.Debug("eth call by timestamp must have the following block hint", zap.String("userName", permsForUser.userName))
-					return http.StatusBadRequest, fmt.Errorf("eth call by timestamp must have the following block hint")
+				call := hex.EncodeToString(callData.Data[0:ETH_CALL_SIG_LENGTH])
+				callKey := fmt.Sprintf("ethCallByTimestamp:%d:%s:%s", pcq.ChainId, contractAddress, call)
+				if _, exists := permsForUser.allowedCalls[callKey]; !exists {
+					logger.Debug("requested call not authorized", zap.String("userName", permsForUser.userName), zap.String("callKey", callKey))
+					return http.StatusBadRequest, fmt.Errorf(`call "%s" not authorized`, callKey)
 				}
+			}
+		case *query.EthCallWithFinalityQueryRequest:
+			for _, callData := range q.CallData {
```

Comment:
Every case is doing the same checks.  Is there a reasonable way to pull them out and share the code?

Reply:
Might be able to create a templated function to do it. Will look into it the next time we add a new query.